### ISSUE TITLE
Data structure

### ISSUE THREAD

**Issue by kailaw138** [Kawue/provim#10]
Thanks again for answering my previous questions, and I have managed to use provim.

I have a question about the data structure (the h5 file) generated, 

Opening the h5 files in HDFView or Vitables reveals the structure of the data: 

![888888](https://github.com/Kawue/provim/assets/119678591/02b8e8a3-9dc0-4134-9019-763bf52234ea)

It can be recognized that axis0 and block0_item is the mz Array, while block0_values is the Data.

Could you clarify which tables are the x Location and y Location? (I guess axis1_label0 and axis1_label1 are x and y)

What are the other tables?

**Comment by Kawue** [comment]
Well, I guess that your assumption is correct.

Short explanation:
The DataFrame uses a MultiIndex meaning that each row is identified by two values (Grid x and Grid y)

![df](https://github.com/Kawue/provim/assets/18120038/e53384a0-a1bd-42c9-a9ef-ac9c41d55607)


Grid y defines the row number and Grid x defines the column number of the m/z image
![mz-image](https://github.com/Kawue/provim/assets/18120038/fd398486-b9a6-47af-9c2a-0d50cc77b87c)


if you want to test which is which, you could try to create an image with the following code:

```
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

df = pd.read_hdf(path-to-file)
data = df[df.columns[any-index-number]].values
x = df.index.get_level_values("grid_x")
y = df.index.get_level_values("grid_y")
w = x.max()
h = y.max()

img = np.zeros((h,w))
img[((y,x))] = data
plt.imshow(img)
plt.show()
```

Additionally you can get your m/z numbers with `df.columns`, the index with `df.index` and a single index, like Grid x with `df.index.get_level_values("grid_x")` or Grid y with `df.index.get_level_values("grid_y")`

With my knowledge today I would structure the data differently. However, it is like it is. This structure as a lot of pros and a lot of cons.

I assume that label and level is the same. So:
- axis1_label0 = axis1_level0 = Grid x
- axis1_label1 = axis1_level1 = Grid y
- axis1_label2 = axis1_level2 = Dataset Name (this might me irrelevant for you as it is only important if you have more than one sample in one dataset. Otherwise it will contain the same entry for the whole dataset)
- axis0 may be equal to block0_items = m/z values
- block0_values are the measured values for each m/z value at each point in the sample


Hope that helps.

**Comment by kailaw138** [comment]
Thank you again for your detailed explanation. I asked these questions because I would like to explore the data with machine learning and neutral network.  

I would like to convert the h5 to a txt file to import to iSegMSI (https://github.com/gankLei-X/iSegMSI) for further data analysis, however, I keep encountering errors because of the "unique file structure" of the ProVim output.

Herewith my code and the error I encounter:
![A](https://github.com/Kawue/provim/assets/119678591/1a0ef621-aa89-47e5-a0af-f0adc356c4cc)
![B](https://github.com/Kawue/provim/assets/119678591/3ddea957-d2d5-40b0-8add-4814b45850b6)

Could you offer me some help?

Alternatively, I am happy to try your [vaidra](https://github.com/Kawue/vaidra), if it does the same. However, there is no documentation, dont know what it does and what to do with it.

and for [grine-v2](https://github.com/Kawue/grine-v2), is there a way not to use docker, but to use windows?

**Comment by Kawue** [comment]
I am currently working on this. I prepare a file for you which explains the difference between h5py and pandas. I guess the conversion should be quite easy. However, I have no suited provim file. I will use an old file that uses another multiiindex. If you provide me a small file processed with provim I could also use this one as example.

Vaidra is only to explore dimension reduction.
Grine could be used without docker but the installation is a lot of work. Why is docker a problem?

If you want to see what these tools are made for I refer to my dissertation: https://doi.org/10.4119/unibi/2956953
Each tools is explained there.

**Comment by Kawue** [comment]
If I understand the format of iSegMSI correctly, you need to use the following:

```
p = "./barley_101.h5" #Path to your file
savepath = "./iSegMSI_data.txt"
data = pd.read_hdf(p)
np.savetxt(savepath, data.values)
```

Here the jupyter notebook file as markdown. Especially the part `# Summarize insights` should be interesting for you. Hope that helps.

[Notebook.md](https://github.com/Kawue/provim/files/14052458/Notebook.md)

**Comment by kailaw138** [comment]
Thank you again for your reply and the code provided, your method using pandas is working. Mine using h5py isnt :( . 

Unfortunately, I have not yet made iSegMSI working yet, 

![i](https://github.com/Kawue/provim/assets/119678591/7081780c-af21-43fb-9767-44f411fa2058)

Perhaps it is because the data is a kidney, and is kidney-shaped. 
(However, using the demo data still does not work and these authors do not respond to my questions. )

I have further questions about VAIDRA.

![v](https://github.com/Kawue/provim/assets/119678591/199b8b58-17ee-45a4-9a01-1c4c693d9921)

After I loaded a h5 file (in this case, only the demo data), it just continued to load, and the windows circle never stopped (indicated by a red arrow). The task manager did not show any workload. 

In case my installation was incorrected, this was how I did the installation:
1. setting up a virtual environment using the environment.yml in ./backend/pyscripts and the activated the environment 
2. setting up the frontend with Node.js 
3. in the activated environment
         cd C:\Program Files\nodejs\
         nodevars.bat
         cd .\VAIDRA\frontend
         set NODE_OPTIONS=--openssl-legacy-provider
         npm install
         npm run build
         npm run serve
4. browser: http://localhost:8080/
5. load h5 ... several hours later ... nothing happened

**Comment by Kawue** [comment]
I do not have that much spare time currently, I will try to have a deeper look into iSegMSI and VAIDRA in the next few days. I guess I will be able to reply to you within this week.

**Comment by Kawue** [comment]
Quick update, I fixed some problems with my vaidra backend but still have a bug in the frontend. After that is solved I will look into the iSegMSI problem.

**Comment by Kawue** [comment]
ok, so if we get vaidra running on your system I will update the repo and add some descriptions.

Please try the following:

- Clone the "Update" Branch (not the main)
- Create a python virtual environment `python -m venv venv_vaidra`
- Activate the virtual environment
- Install the `requirements.txt` which is in the backend folder into the virtual environment with `pip install -r requirements.txt`
- Start the backend `python .\api.py`
- Open a second Powershell or CMD and navigate into the frontend folder
- Use `yarn install` within the frontend folder, there might be errors, ignore them
- Use `yarn serve` to start the frontend, there might be some more errors, ignore these again
- Vaidra should be available under `127.0.0.1:8080`
- In Vaidra use the + to open the side drawer, select your data set, the dimensions and the methods (I recommend PCA & UMAP for first tests), then press the run button. Errors in processing will occur in the powershell that started the `api.py`
- Now there is a small bug within the workflow. After the processing is done, stop the backend (CTRL+C) and run it again.
- Now you can select an embedding, and both dataset dropdowns. After that you have to select an x and y dimension.
- A plot should be shown. Using the "Lasso" button allows to select some points from the plot.

If all this works I can give you a more detailed introduction into the tool.

**Comment by kailaw138** [comment]
thanks again

i got as far as step 7: Use yarn install within the frontend folder

![2](https://github.com/Kawue/provim/assets/119678591/9626f3f3-d764-4a0e-b62a-2512630c4095)

error @achrinza/node-ipc@9.2.2: The engine "node" is incompatible with this module. Expected version "8 || 10 || 12 || 14 || 16 || 17". Got "20.10.0"
error Found incompatible module.
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.

may i suggest that it would be easier (to let users like me) to have a series of docker images readily available (for the provim family) and just need to download and run (no need to build or set up environment). what do u think?

**Comment by Kawue** [comment]
Well of course this would be way easier. The major problem is that I do not work at university anymore and all support I provide currently is only in my spare time. If I have time left I can try to set up a container which is ready to use but currently my daily job is keeping me quite busy.

You can try to use `yarn install --ignore-engines` to solve this issue.


I also got iSegMSI working. I will provide you the code that I used to format the data correctly as soon as I cleaned it. Its quite messy currently.

**Comment by Kawue** [comment]
So no progress with putting Vaidra into a docker so far. However, its still on my todo list.

Here is the Code with which I converted my example file into the correct format for iSegMSI

```
import pandas as pd
import os

p = "./barley_101.h5" # CHANGE PATH HERE
d = pd.read_hdf(p)

grid_x = d.index.get_level_values("grid_x")
grid_y = d.index.get_level_values("grid_y")
all_pos_xy = list(product(range(max(grid_x)+1), range(max(grid_y)+1)))
pixel_index = list(zip(grid_x, grid_y))
empty_pixels = list(set(all_pos_xy) - set(pixel_index))
dset_name = d.index.get_level_values("dataset")[0]
empty_pixels = [(x[0], x[1], dset_name) for x in empty_pixels]
empty_pixels_index = pd.MultiIndex.from_tuples(empty_pixels, names=["grid_x", "grid_y", "dataset"])
empty_df = pd.DataFrame(0, index=empty_pixels_index, columns=d.columns)
full_df = pd.concat([d,empty_df])
full_df = full_df.sort_index(level=0) #1
full_df = full_df.sort_index(level=1) #0

s_path = os.path.normpath("./iSegMSI_format")
if not os.path.exists(s_path):
    os.mkdir(s_path)
full_df.to_hdf(os.path.join(s_path, dset_name + ".h5"), key=dset_name, complib="blosc", complevel=9)

savepath = "./iSegMSI_data.txt"
p = "./iSegMSI_format/barley_101.h5"
data = pd.read_hdf(p)
np.savetxt(savepath, data.values)
```

Then I ran
```
python run.py --input_file ../iSegMSI_data.txt --input_shape 52 111 101 --DR_mode umap --n_components 3 --use_scribble 0 --output_file output_barley.txt
```

Works also with `hyperparameter_search.py` instead of `run.py`

For the scribble approach I currently have no idea how the scribble file has to be created. If you have access to the paper, I would like to read it.

The results seem promising. At least for my well structured barley seed I can find all relevant morphological structures within the segmentation map.

EDIT: I forgot that I had to change line 137 in model.py of the iSegMSI project. I replaced `107` with `args.input_shape[1]`.

**Comment by kailaw138** [comment]
Thank you again for your support. Apologize for late reply. It was new year holiday in this side of the world and I was away from the office. 

the updated version of vaidra works

![3](https://github.com/Kawue/provim/assets/119678591/e33d3259-0237-43eb-8d4d-b835064e7332)

However, I do not know how to get projection images. More instructions would be appreciated.

**Comment by kailaw138** [comment]
As for the python code to convert provim output to iSegMSI input, I encountered problems.

![hh](https://github.com/Kawue/provim/assets/119678591/3b45ad64-eb6a-44ae-8b73-1574034b3f1e)

so i define "product"

![hhhh](https://github.com/Kawue/provim/assets/119678591/cd67db6c-13db-4dd4-815c-46500ca2c1ad)

(Please let me have your email address, I forward the articles to you)

**Comment by kailaw138** [comment]
updated: as for the code to convert provim output to iSegMSI input

I (actually it was Gemini) made some modifications to get it work...

```
import pandas as pd
import os
import numpy as np  ###
from itertools import product  ###

p = "G:/Kidney/DAN_neg_ProViM/Output/20231110_rat_kidney_DAN_100um_neg_FI_processed_simplified_cleaned_pruned.h5" # CHANGE PATH HERE
d = pd.read_hdf(p)

grid_x = d.index.get_level_values("grid_x")
grid_y = d.index.get_level_values("grid_y")
all_pos_xy = list(product(range(max(grid_x)+1), range(max(grid_y)+1)))
pixel_index = list(zip(grid_x, grid_y))
empty_pixels = list(set(all_pos_xy) - set(pixel_index))
dset_name = d.index.get_level_values("dataset")[0]
empty_pixels = [(x[0], x[1], dset_name) for x in empty_pixels]
empty_pixels_index = pd.MultiIndex.from_tuples(empty_pixels, names=["grid_x", "grid_y", "dataset"])
empty_df = pd.DataFrame(0, index=empty_pixels_index, columns=d.columns)
full_df = pd.concat([d,empty_df])
full_df = full_df.sort_index(level=0) #1
full_df = full_df.sort_index(level=1) #0

s_path = os.path.normpath("G:/Kidney/DAN_neg_ProViM/Output/iSegMSI_format")
if not os.path.exists(s_path):
    os.mkdir(s_path)
full_df.to_hdf(os.path.join(s_path, dset_name + ".h5"), key=dset_name, complib="blosc", complevel=9)

savepath = "G:/Kidney/DAN_neg_ProViM/Output/iSegMSI_data.txt"
data = pd.read_hdf(os.path.join(s_path, dset_name + ".h5"))  ###
np.savetxt(savepath, data.values)
```

Well, the output runs with iSegMSI, but the segmentation looks awful.

![iSegMSI_unsupervised_output txt](https://github.com/Kawue/provim/assets/119678591/f5566bde-2629-424f-8054-29e68c677be5)

**Comment by Kawue** [comment]
Regarding VAIDRA there seems to be a bug. The Dimension Reduced images should appear in the "card views" below dimension X and dimension Y. I currently work on a Docker version but have some bugs left. I will notify you as soon as it runs. Does your browser inspector give you any error messages? It seems like it cant access the images.

Regarding iSegMSI I assume that you made a mistake with `--input_shape`. Your image is not really the segmentation. Its the result of a pixel arrangement error. I had the same problem. Please provide me information about the following:

+ Value of `grid_x`
+ Value of `grid_y`
+ Value of `len(d.columns)`
+ The three parameters after `--input_shape`

**Comment by kailaw138** [comment]
Value of grid_x  0-102
Value of grid_y  0-183
Value of len          0-7999 [m/z 99.951-1993.587]

so I set --input_shape 103 184 8000

the file 20231110_rat_kidney_DAN_100um_neg_FI_processed_simplified_cleaned_pruned.h5，can be downloaded from: https://pan.quark.cn/s/7b80294eb13b

________________________________

**Comment by Kawue** [comment]
Would it be possible for you to share it with a service that does not require a login and uses english?

I know that for some data that may be problematic, but it would make it much easier for me to access. If you do not trust any provider, then onionshare would be an option.

Its always easier if I have the option to play with the data myself.

**Comment by kailaw138** [comment]
I am afraid that Tor is being blocked...

try google:

https://drive.google.com/file/d/1oIJMy1gAYj817wD-MWwn6mysxFTUFaLl/view?usp=sharing
________________________________

**Comment by Kawue** [comment]
I got the following result

![output txt](https://github.com/Kawue/provim/assets/18120038/b7de9feb-9192-43d6-9d8d-7a16f8685753)

To achieve this you have to change
```
full_df = full_df.sort_index(level=0) 
full_df = full_df.sort_index(level=1) 
```

to

```
full_df = full_df.sort_index(level=1) 
full_df = full_df.sort_index(level=0) 
```

So just interchange 0 and 1. This has to do with the orientation of your sample.

The call was:
`python run.py --input_file ../iSegMSI_format_china.txt --input_shape 103 184 8000 --DR_mode umap --n_components 3 --use_scribble 0 --output_file output`

Alternatively you can run the `hyperparameter_search.py`

**Comment by kailaw138** [comment]
Thanks, eventually work like charm.

I have forwarded you the articles via researchgate.

The scribbles method according to the paper "Using prior information in the form of scribbles, the unsupervised model is regularized to correct inappropriate segments specified by the imputing scribbles. Briefly, an unsupervised segmentation model is first constructed. Then, a two-phase scheme is proposed to train the model. At the first phase, the unsupervised model is pretrained to achieve a preliminary segmentation result. At the second phase, handdrawn scribbles are used to impose regularizations on the specified inappropriate segmented regions. In short, the model is fine-tuned to achieve more accurate segmentation results."

It requires a rough (histology) drawing (supp. figure s2a) as a guide. The script to convert the drawing to a scribble.txt is not disclosed.


________________________________

**Comment by kailaw138** [comment]
update:

well, taken several hours to run the hyperparameter_search.py, but only txt files are produced. The image file is being overwritten each round, only the last one is kept.
________________________________

**Comment by Kawue** [comment]
As far as I understand the idea the hyperparameter search optimizes the parameters in each step. Therefore the intermediate images are overwritten and only the last "optimal" result is used as output. I would prefer to create a single image for each step but I think thats preference.

Although iSegMSI works for you I would like to keep this ticket open. I still work on the docker version of VAIDRA to make it usable. I have a feeling that it might be interesting for you to explore your data with different dimension reduction results.

I would also like to mention that based on my experience, the first few dimension reduction images are "trivial".
Dimension 0/1 are often a differentiation between background and foreground, which basically means matrix and sample molecules for MSI. The next few dimensions are often big morphological regions that differ in the molecular composition. Something like malignant tissue is often somewhere in the later dimensions.

I will keep you updated once VAIDRA works.

Feel free to write any further questions in this ticket.

p.s. thanks for the documents.
